### PR TITLE
doc/website/blog: versioned migration authoring blog post

### PR DESCRIPTION
**Pull request by rotemtam**

### CODE REVIEW

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 1]
```diff
@@ -0,0 +1,243 @@
+---
+title: Announcing v0.6.0 with Versioned Migration Authoring
+authors: a8m
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:3]:
```suggestion
authors: rotemtam
```
Not?

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+tags: [atlas, migrations, versioned]
+image: https://blog.ariga.io/uploads/images/posts/v0.5.0/atlas-lint.png
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:5]: Make a new image

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+
+With the release of [v0.6.0](https://github.com/ariga/atlas/releases/tag/v0.6.0), we are
+happy to announce official support for a style of workflow for managing changes to database
+schemas that we have been experimenting with in the past months: _Versioned Migration Authoring_.
+
+### tl;dr
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:16]:
```suggestion
### TL;DR
```
Or `TLDR`

I am not sure I like this section. It breaks the reading flow because the post reads repetitive since we read parts doubly.

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+* Atlas supports a declarative workflow (similar to Terraform) where users provide the 
+ desired database schema in a simple [data definition language](https://atlasgo.io/atlas-schema/sql-resources)
+ and Atlas calculates a plan to get a target database to that state. This workflow
+ is supported by the [`schema apply`](https://atlasgo.io/declarative/apply) command.
+* Many teams prefer a more imperative approach where each change to the database schema 
+ is checked-in to source control and reviewed during code-review. This type of workflow
+ is commonly called _versioned migrations_ and is supported by many established tools such
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:24]: Especially folks coming from Java Spring will know the term change based migrations. Maybe we add that term as well to decrease friction for readers from there?

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  schema = schema.main
+  column "id" {
+    type = int
+  }
+  column "greeting" {
+    type = text
+  }
+}
+```
+Now, suppose that you want to add a default value of `"shalom"` to the `greeting`
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:64]:
```suggestion

Now, suppose that you want to add a default value of `"shalom"` to the `greeting`
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+-- Rename temporary table "new_users" to "users"
+ALTER TABLE `new_users` RENAME TO `users`
+```
+
+### Versioned Migrations
+
+As the database is one of the most critical components in any system, applying changes
+to its schema is rightfully considered a dangerous operation. For this reason, many teams
+prefer a more imperative approach where each change to the database schema is checked-in
+to source control and reviewed during code-review.  Each such change
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:103]:
```suggestion
to source control and reviewed during code-review. Each such change
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+is called a "migration", as it migrates the database schema from the previous version to
+the next. To support this kind of requirement, many popular database schema management
+tools such as [Flyway](https://flywaydb.org/), [Liquibase](https://liquibase.org/) or
+[golang-migrate](https://github.com/golang-migrate/migrate) support a workflow that
+is commonly called "versioned migrations".
+
+In addition to the higher level of control which is provided by versioned migrations,
+applications are often deployed to multiple remote environments at once. These environments,
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:111]:
```suggestion
applications are often deployed to multiple remote environments at once. These environments
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+interpret the migration files and to apply (some of) them in the correct order to
+transition to the desired database structure.
+
+The benefit of the versioned migrations approach is that it is explicit: engineers
+know _exactly_ what queries are going to be run against the database when the time
+comes to execute them.  Because changes are planned ahead of time, migration authors
+can control precisely how to reach the desired schema.  If we consider a migration as
+a plan to get from state A to state B, oftentimes multiple paths exist, each with a
+very different impact on the database. To demonstrate, consider an initial state which
+contains a table with two columns:
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:130]:
```suggestion
contains a table with two columns:

```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+into normal migration files which can be checked into source control, fine-tuned manually and
+reviewed in regular code review processes.
+
+### Getting started
+
+Start by downloading the Atlas CLI. The following command will work for MacOS:
+```bash
+brew install ariga/tap/atlas
+```
+For instructions on other platforms [go here](cli/getting-started/setting-up#install-the-cli).
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:177]: Lets copy the tab from the install page to have instructions for the three major platforms? If I wanna follow along and have to jump somwhere else mid-read it disrupts the flow.

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+Let's run `atlas migrate diff` with the necessary parameters to generate a migration script for
+creating our `users` table:
+
+* `--dir` the URL to the migration directory, by default it is `file://migrations`.
+* `--to` the URL of the desired state, an HCL file or a database connection.
+* `--dev-url` a [URL](/concepts/url) to a [Dev Database](/concepts/dev-database) that will be used to compute the diff.
+
+```bash
+atlas migrate diff --dir="file://migrations" --to="file://schema.hcl" \
+ --dev-url="mysql://root:pass@:3306/test" create_users
```
Comment (masseelch) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:201]:
```suggestion
atlas migrate diff create_users \
  --dir="file://migrations" \
  --to="file://schema.hcl" \
  --dev-url="mysql://root:pass@:3306/test"
```

**Review by masseelch — COMMENTED** [review]
Great start :rocket:

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx (hunk 2 above)
Comment (rotemtam) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:5]:
```suggestion
image: https://blog.ariga.io/uploads/images/posts/v0.6.0/atlas-migrate-diff.png
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx (hunk 4 above)
Comment (rotemtam) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:24]:
```suggestion
 is commonly called _versioned migrations_ (or _change based migrations_) and is supported by many established tools such
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx (hunk 9 above)
Comment (rotemtam) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:177]:
```suggestion

<Tabs
defaultValue="apple-intel-brew"
values={[
{label: 'macOS (Homebrew)', value: 'apple-intel-brew'},
{label: 'macOS (Plain)', value: 'apple-intel-plain'},
{label: 'Linux', value: 'linux'},
{label: 'Windows', value: 'windows'},
]}>
<TabItem value="apple-intel-brew">

Get the latest release with [Homebrew](https://brew.sh/):

```shell
brew install ariga/tap/atlas
```

</TabItem>
<TabItem value="apple-intel-plain">

Download latest release.
```shell
curl -LO https://release.ariga.io/atlas/atlas-darwin-amd64-latest
```

Make the atlas binary executable.
```shell
chmod +x ./atlas-darwin-amd64-latest
```

Move the atlas binary to a file location on your system PATH.
```shell
sudo mv ./atlas-darwin-amd64-latest /usr/local/bin/atlas
```
```shell
sudo chown root: /usr/local/bin/atlas
```

</TabItem>
<TabItem value="linux">

Download latest release.
```shell
curl -LO https://release.ariga.io/atlas/atlas-linux-amd64-latest
```

Move the atlas binary to a file location on your system PATH.
```shell
sudo install -o root -g root -m 0755 ./atlas-linux-amd64-latest /usr/local/bin/atlas
```

</TabItem>
<TabItem value="windows">

Download the [latest release](https://release.ariga.io/atlas/atlas-windows-amd64-latest.exe) and
move the atlas binary to a file location on your system PATH.


</TabItem>
</Tabs>
```

**Comment by masseelch** [comment]
https://giphy.com/gifs/gutearbeitoriginals-perfect-top-l3q2LH45XElELRzRm

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+tags: [atlas, migrations, versioned]
+image: https://blog.ariga.io/uploads/images/posts/v0.6.0/atlas-migrate-diff.png
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+
+With the release of [v0.6.0](https://github.com/ariga/atlas/releases/tag/v0.6.0), we are
+happy to announce official support for a style of workflow for managing changes to database
```
Comment (hilakashai) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:13]:
```suggestion
happy to announce official support for a workflow style for managing changes to database
```

Maybe? idk something about this sounds weird 🤔

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+DROP TABLE `users`
+-- Rename temporary table "new_users" to "users"
+ALTER TABLE `new_users` RENAME TO `users`
+```
+
+### Versioned Migrations
+
+As the database is one of the most critical components in any system, applying changes
+to its schema is rightfully considered a dangerous operation. For this reason, many teams
+prefer a more imperative approach where each change to the database schema is checked-in
```
Comment (hilakashai) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:103]:
```suggestion
prefer a more imperative approach where each change to the database schema is checked in
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+
+### TL;DR
+
+* Atlas supports a declarative workflow (similar to Terraform) where users provide the 
+ desired database schema in a simple [data definition language](https://atlasgo.io/atlas-schema/sql-resources)
+ and Atlas calculates a plan to get a target database to that state. This workflow
+ is supported by the [`schema apply`](https://atlasgo.io/declarative/apply) command.
+* Many teams prefer a more imperative approach where each change to the database schema 
+ is checked-in to source control and reviewed during code-review. This type of workflow
+ is commonly called _versioned migrations_ (or _change based migrations_) and is supported by many established tools such
```
Comment (hilakashai) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:24]:
```suggestion
 is commonly called _versioned migrations_ (or _change-based migrations_) and is supported by many established tools such
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+[golang-migrate](https://github.com/golang-migrate/migrate) support a workflow that
+is commonly called "versioned migrations".
+
+In addition to the higher level of control which is provided by versioned migrations,
+applications are often deployed to multiple remote environments at once. These environments
+are not controlled (or even accessible) by the development team. In such cases, declarative migrations,
+which rely on a network connection to the target database and on human
+approval of migrations plans in real-time, are not a feasible strategy.
+
+With versioned migrations (sometimes called "change-based migrations") instead of describing
```
Comment (hilakashai) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:117]:
```suggestion
With versioned migrations (sometimes called "change-based migrations"), instead of describing
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+    user_name varchar(255)
+);
+```
+There are at least two ways get from the initial to the desired state:
+* Drop the `name` column and create a new `user_name` column.
+* Alter the name of the `name` column to `user_name`.
+
+Depending on the context, either may be the desired outcome for the developer
+planning the change. With versioned migrations, engineers have the ultimate confidence
+of what change is going to happen which may not be known ahead of time in a _declarative_
```
Comment (hilakashai) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:152]:
```suggestion
of what change is going to happen, which may not be known ahead of time in a _declarative_
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+* `--dev-url` a [URL](/concepts/url) to a [Dev Database](/concepts/dev-database) that will be used to compute the diff.
+
+```bash
+atlas migrate diff create_users \
+  --dir="file://migrations" \
+  --to="file://schema.hcl" \
+  --dev-url="mysql://root:pass@:3306/test"
+```
+
+Observe two files were created in the `migrations` directory:
```
Comment (hilakashai) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:264]:
```suggestion
Observe that two files were created in the `migrations` directory:
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+-- create "users" table
+CREATE TABLE `users` (`id` int NOT NULL) CHARSET utf8mb4 COLLATE utf8mb4_0900_ai_ci;
+```
+
+</TabItem>
+<TabItem value="sum_file">
+
+In addition to the migration directory, Atlas maintains a file name `atlas.sum` which is used
+to ensure the integrity of the migration directory and force developers to deal with situations
+where migration order or contents was modified after the fact.
```
Comment (hilakashai) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:287]:
```suggestion
where migration order or contents were modified after the fact.
```

**Review by hilakashai — COMMENTED** [review]
💯

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx (hunk 14 above)
Comment (a8m) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:117]:
```suggestion
With versioned migrations, instead of describing
```
It's already mentioned above. It feels unnecessary to mention it again.

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+the desired state ("what the database should look like"), developers describe the changes themselves
+("how to reach the state"). Most of the time, this is done by creating a set of SQL files
+containing the statements needed. Each of the files is assigned a unique version and a
+description of the changes. Tools like the ones mentioned earlier are then able to
+interpret the migration files and to apply (some of) them in the correct order to
+transition to the desired database structure.
+
+The benefit of the versioned migrations approach is that it is explicit: engineers
+know _exactly_ what queries are going to be run against the database when the time
+comes to execute them.  Because changes are planned ahead of time, migration authors
```
Comment (a8m) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:127]:
```suggestion
comes to execute them. Because changes are planned ahead of time, migration authors
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+can control precisely how to reach the desired schema.  If we consider a migration as
```
Comment (a8m) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:128]:
```suggestion
can control precisely how to reach the desired schema. If we consider a migration as
```

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+The downside of the _versioned migration_ approach is, of course, that it puts the
+burden of planning the migration on developers. This requires a certain level
+of expertise that is not always available to every engineer, as we demonstrated
+in our example of setting a default value in a SQLite database above.
+
+As part of the Atlas project we advocate for a third combined approach that we call
+"Versioned Migration Authoring". Versioned Migration Authoring is an attempt to combine
+the simplicity and expressiveness of the declarative approach with the control and
+explicitness of versioned migrations.
+
```
Comment (a8m) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:166]: 💯

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 21]
```diff
@@ -0,0 +1,303 @@
...
+the Atlas engine to plan a safe migration from the existing to the new state.
+However, instead of coupling planning and execution, plans are instead written
+into normal migration files which can be checked into source control, fine-tuned manually and
+reviewed in regular code review processes.
+
+### Getting started
+
+Start by downloading the Atlas CLI:
+
+<Tabs
```
Comment (a8m) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:177]: Idea: create a react component and embed it every time we mention the getting-started.

File: doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx [hunk 22]
```diff
@@ -0,0 +1,303 @@
...
+</TabItem>
+</Tabs>
+
+### Further reading
+
+To learn more about Versioned Migration Authoring:
+* Read the [docs](/versioned/diff)
+* [CLI Command Reference](/cli-reference#atlas-migrate-diff)
+
+Have questions? Feedback? Find our team [on our Discord server](https://discord.gg/zZ6sWVg6NT).
```
Comment (a8m) [inline doc/website/blog/2022-08-11-announcing-versioned-migration-authoring.mdx:303]: Excellent

**Review by a8m — APPROVED** [review]
Fantastic 💯
